### PR TITLE
Calendar backend sync

### LINKED ISSUE #27: creating appointments
At the beginning of a new semester, a VLA member clicks somewhere (tbd) in the calendar, and a page to create a new appointment opens. This form has fields to enter 
* the title
* the lecture type it belongs to (optional) (#34)
* category (lecture/vacation/Abnahmetermin/...) (#28)
* start datetime, end datetime, recurrence week days, recurrence end date,
* peoples / lecturers: there may be multiple people (Prof, Assistenz) per appointment (#33)

After submitting the form, this appointment is visible in the main calendar view, colored according to its category.

---

Questions to be answered:

* [x] person: list of existing people or free-text form?

ToDo:

* [x] lectures -> #69 
* [x] appointment categories -> #70 
* [x] persons -> #71 
* [ ] lecture_persons
* [x] appointment_series -> #99 
* [x] appointments -> #99 
* [x] acceptance -> #99 
* [x] experiment_booking -> #99

### LINKED ISSUE #28: creating appointment categories
During first use of this application, a VLA member opens the page for appointment categories. A list of existing categories is displayed, each category has a color. The VLA member clicks a button to create a new category, enters a name and picks a color. After submitting, the list of categories is shown, where the newly created category has been added. Later during appointment creation (#27), I can choose out of all these categories.

### LINKED ISSUE #33: creating people
As a VLA member, I click somewhere (tbd) in this application to create a new person which is holding or assisting future lectures or a new VLA member. A form shows up where I can enter their name. After submitting, this person is shown as one proposal for associated people upon appointment creation (#27).

Things to determine: Whether these "personas" are a separate table next to the "users" (which are able to log in into our application), or we allow user creation w/o setting username/password (and thus not allowing them to login theirselves).

### LINKED ISSUE #34: creating lectures
As a VLA member, I love Physik 1 and create this as a lecture type in some form (tbd). The form allows me to enter the name of this lecture. After submitting this form, this lecture is proposed during appointment creation (#27)

### LINKED ISSUE #127: frontend: adapt logic of event creation w.r.t. series if no lecture specified
To have smooth behaviour for moving experiments (e#19 ), events without lecture should not be in a separate series per weekday.

This requires a very small logic change in handleCreateEvent function of useEvents.ts in the frontend.

### PR DESCRIPTION
**Pull request by JohaJung**
Finally makes everything persistent.
This replaces the browserwindow-internal `useState`s with an SSE-bound state, as introduced by @MarcoLeipnitz in the `useSseWithInitialFetch` helper.

Currently, appointment categories and people are synced. Before merging, I plan to also include

* [x] Lectures
* [x] Appointments + Appointment series

closes #33
closes #34
closes #27
closes #28 
closes #127 

---

### Implementation

- [x] Does this code change accomplish what it is supposed to do?
- [x] Can this solution be simplified without unreasonable effort
- [x] Is the code modular enough?
- [x] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [x] Single Responsibility Principle: Does any class serve only one purpose? If not, it should probably split up in multiple classes.
- [x] Interface Segregation: If interface implemented, is it fully implemented?
- [x] Dependency Injection: Are central object instantiation processes managed by a single context?
- [x] If any functionality is missing, is it marked with a comment that starts with TODO:?
- [x] Do you see any potential to improve the performance of the code significantly? 

### Logic Errors and Bugs

- [x] Can you think of any use case in which the code does not behave as intended?
- [x] Can you think of any inputs or external events that could break the code?
- [ ] Are edge cases covered by tests?

### Error Handling and Logging

- [x] Is error handling done the correct way?
- [x] Should any logging or debugging information be added or removed?
- [x] Are error messages user-friendly?

### Dependencies (internal and external)

- [x] Were updates to documentation, configuration, or readme files made as required by this change?
- [x] Is the API.md up-to-date?
- [x] Are the enumerations between frontend and backend synchronized?

### Security and Data Privacy

- [x] Does the code introduce any security vulnerabilities?
- [x] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [x] Is sensitive data like user data handled and stored securely?
- [x] Is the right encryption used?
- [x] Is data retrieved from external APIs or libraries checked for security issues?
- [x] Were any secrets committed?

### Testing and Testability (later)

- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Do the existing tests reasonably cover the code change?
- [ ] Are there some test cases, input or edge cases that should be tested in addition?

### Usability and Accessibility

- [x] Is the proposed solution (UI) accessible (e.g. colorblindness, ...)?
- [x] Is the API/UI intuitive to use?
- [ ] Was the app tested on another device?

### Readability

- [x] Is the code easy to understand?
- [x] Which parts were confusing to you and why?
- [x] Can the readability of the code be improved by smaller methods/different names?
- [x] Is the code located in the right file/folder?
- [x] Do you think certain methods should be restructured to have a more intuitive control flow?
- [x] Are there redundant or outdated comments?
- [x] Could some comments convey the message better?
- [x] Could some comments be removed by making the code itself more readable?

### CODE REVIEW

File: VLA-Frontend/src/hooks/usePeople.ts [hunk 1]
```diff
@@ -1,28 +1,58 @@
...
+  const [people, _setPeople] = useSseConnectionWithInitialFetch<Person[]>(
+    [], API_URL, sseHandlers
+  );
 
-  function handleAddPerson(person: Person) {
-    //TODO: Backend - POST request to /api/people
-    setPeople((prev) => [...prev, person]);
+  async function handleAddPerson(person: Person) {
+    return fetch(API_URL, {
+      method: "POST",
```
Comment (polodasi123) [inline VLA-Frontend/src/hooks/usePeople.ts]: hier auch

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java [hunk 2]
```diff
@@ -91,8 +121,7 @@ public ResponseEntity<?> deleteAppointment(@PathVariable Long id) {
             () -> new EntityNotFoundException(
                 "Appointment with ID " + id + " not found."));
         appointmentRepository.deleteById(id);
-        // TODO: use a better method here instead of debug message
-        SseController.notifyDebugTest("Appointment deleted: " + deletedAppointment);
+        SseController.notifyAllOfObject(SseMessageType.APPOINTMENTDELETED, deletedAppointment);
         return ResponseEntity.ok(deletedAppointment);
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java]: Off-topic, aber klingt sinnvoll.

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 3]
```diff
@@ -1,16 +1,52 @@
...
   getEventDateISO,
   moveEventSeries,
   verifyValidTimeRange
 } from "@/components/calendar/eventUtils.ts";
 import {Logger} from "@/components/logger/Logger.ts";
 import {getNotSynchronisedId} from "@/lib/utils.ts";
+import useSseConnectionWithInitialFetch from "@/hooks/useSseConnectionWithInitialFetch.ts";
 
+const API_URL_APPOINTMENTS = "/api/appointments";
+const API_URL_SERIES = "/api/appointmentSeries";
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/hooks/useEvents.ts]: Sagt der, der in den GlobalNotes-Dateien **viermal** den String "/api/globalNotes" hardcodet hat :stuck_out_tongue_winking_eye:

File: VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java [hunk 4]
```diff
@@ -61,6 +72,8 @@ public ResponseEntity<?> createAppointment(@RequestBody Appointment appointment)
     @PutMapping("/{id}")
     public ResponseEntity<?> updateAppointment(@PathVariable Long id,
```
Comment (MarcoLeipnitz) [inline VLA-Backend/src/main/java/de/vlaorgatu/vlabackend/controller/vladb/AppointmentController.java]: Make generic narrower

File: VLA-Frontend/src/components/globalNotes/GlobalNoteContainer.tsx [hunk 5]
```diff
@@ -40,7 +41,7 @@ export default function GlobalNoteContainer() {
   const [globalNotes, _setGlobalNotes] = 
     useSseConnectionWithInitialFetch<GlobalNote[]>(
       [],
-      "/api/globalNotes",
+      API_URL_GLOBAL_NOTES,
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/components/globalNotes/GlobalNoteContainer.tsx]: Thank you :)

File: VLA-Frontend/src/hooks/useCategories.ts [hunk 6]
```diff
@@ -1,19 +1,61 @@
...
+import {API_URL_APPOINTMENT_CATEGORIES} from "@/lib/api.ts";
+import {Logger} from "@/components/logger/Logger.ts";
+
+function handleCategoryCreated(event: MessageEvent, currentValue: AppointmentCategory[]) {
+  const newCategory = JSON.parse(event.data) as AppointmentCategory;
+  return [...currentValue, newCategory];
+}
+
+function handleCategoryDeleted(event: MessageEvent, currentValue: AppointmentCategory[]) {
+  const deletedNote = JSON.parse(event.data) as AppointmentCategory;
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useCategories.ts]: outdated variable name

File: VLA-Frontend/src/hooks/useCategories.ts [hunk 7]
```diff
@@ -1,19 +1,61 @@
...
+  return currentValue.filter((cat) => cat.id !== deletedNote.id);
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useCategories.ts]: still not a note

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 8]
```diff
@@ -1,35 +1,95 @@
...
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: getJSONEvent({...event, notes: notes}),
+      }).then((response) => {
+        if (!response.ok) {
+          throw new Error("Error during appointment update: " + response.statusText + ".");
+        }
+      }).catch((error) => {
+        Logger.error("Error after updating event notes: " + error);
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useEvents.ts]: console.log it as well

File: VLA-Frontend/src/hooks/useLectures.ts [hunk 9]
```diff
@@ -1,32 +1,66 @@
...
   /**
-   * Remove a lecture by id.
+   * Remove a lecture.
    */
-  function handleDeleteLecture(lecture: Lecture) {
-    setLectures((prev) => prev.filter((prevLecture) => prevLecture.id !== lecture.id));
+  async function handleDeleteLecture(lecture: Lecture) {
+    return fetch(`${API_URL_LECTURES}/${lecture.id}`, {
+      method: "DELETE",
+    }).then((response) => response.json()).then((deletedLecture) => deletedLecture as Lecture);
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useLectures.ts]: Please give me a \n.

File: VLA-Frontend/src/hooks/useSseConnectionWithInitialFetch.ts [hunk 10]
```diff
@@ -5,15 +5,20 @@ import type { SseMessageType } from "@/lib/databaseTypes";
...
+ * @param jsonReviver A function that will be given to JSON.parse as reviver parameter.
  * @returns The reactive variable
  */
 export default function useSseConnectionWithInitialFetch<T extends object>(
   defaultValue: T, 
   apiResourceURL : string,
-  eventHandlers: Map<SseMessageType, (event: MessageEvent, value: T) => T>
+  eventHandlers: Map<SseMessageType, (event: MessageEvent, value: T) => T>,
+  // explicit any here, because what else should be the type annotation?
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/useSseConnectionWithInitialFetch.ts]: Consider turning the no-explicit-any rule off in the eslint config
Disabling it always does not make sense

File: VLA-Frontend/src/hooks/useEvents.ts (hunk 8 above)
Comment (JohaJung) [reply VLA-Frontend/src/hooks/useEvents.ts]: No, we should do this in the central Logger. I like the Logger class to be the central logging handler.

Especially if removing the component in the UI, we should just change the logic inside Logger to console.log everything. I definitely don't want to adjust all those callsites just because we removed a debug-log button in the UI.

File: VLA-Frontend/src/hooks/useEvents.ts [hunk 11]
```diff
@@ -151,26 +318,42 @@ export function useEvents() {
...
-    // TODO: Backend - POST request to /api/appointments
-    // Append newly created events to state
-    setEvents((prev) => [...prev, ...newEvents]);
+    try {
+      const response = await fetch(`${API_URL_APPOINTMENTS}/multi`, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+        },
+        body: "[" + newEvents.map(e => getJSONEvent(e)).join(",") + "]",
```
[the remark this answers is not in the document]
Comment (JohaJung) [reply VLA-Frontend/src/hooks/useEvents.ts]: Solved with the new central wrapper.

File: VLA-Frontend/src/hooks/useSseConnectionWithInitialFetch.ts (hunk 10 above)
Comment (JohaJung) [reply VLA-Frontend/src/hooks/useSseConnectionWithInitialFetch.ts]: I did not research how to disable, and I consider it a good pattern to have this rule in general. There are just those two special cases where the general JSON method signature forces us to give an any function.

In our own code, I'd like to avoid any typing, still.

**Comment by JohaJung** [comment]
>* The ganztägig Button:
>       
>   * It sets both start and end time to 23:59
>   * If no time is chosen yet, it does nothing (maybe intended)
>   * It is possible to create appointments with ordinary times and ganztägig ticked (maybe intended, but weird)

fixed, i forgot a return in setState call :see_no_evil: 

>     * Editing Series:
>       
>       * Changing lecture is possible, but not to no lecture (cant be picked in menu)

fixed

**Comment by MarcoLeipnitz** [comment]
There were outdated field names in the helper method..

File: VLA-Frontend/src/hooks/usePeople.ts [hunk 12]
```diff
@@ -1,28 +1,61 @@
...
-    setPeople((prev) =>
-      prev.map((prevPerson) =>
-        prevPerson.id === personId ? { ...prevPerson, notes: notes } : prevPerson
-      )
-    );
-    console.log("new notes: " + people.find((person) => person.id === personId)?.notes);
+  /**
+   * Update the notes of a person.
+   */
+  function handleUpdatePersonNotes(personId: number, notes: string): void {
```
Comment (MarcoLeipnitz) [inline VLA-Frontend/src/hooks/usePeople.ts]: void return type?

**Comment by JohaJung** [comment]
> please adapt the series creation with lectures, so that we can go forward with the decision tree for Appointment Deletion you provided internally.

Done. 

One can torpede this behaviour by adding / removing a lecture from an appointment series afterwards, of course... But this is not a problem for this PR.
